README.md
# coinchart

#### Display cryptocurrency candlestick charts in your terminal.

![screenshot](https://user-images.githubusercontent.com/8098110/219873326-f8c39078-e1f5-4485-97fb-803dd2d1c4fd.png)


## Installation

```bash
npm i -g coinchart
```

## Examples 


```bash
coinchart
```

Returns the default ie. BTC/USDT 1h interval chart.

```bash
coinchart eth
```

Returns ETH/USDT 1h interval chart.

```bash
coinchart ada -i 1d
```

Returns ADA/USDT 1d interval chart.

## Options

```bash
Options:
  -h, --help      Show help
  -i, --interval  Interval eg. 15m, 4h, 1d                       [default: "1h"]
  -p, --pair      Coin pairing eg. BTC, BNB, ETH               [default: "USDT"]
  -l, --limit     Number of candlesticks                           [default: 50]
  -r, --rows      Graph height in rows                             [default: 25]
  -v, --version   Show version number
```

## Contribute

Feel free to make a PR with any improvements or features.

## Acknowledgements

Thanks to [asciichart](https://www.npmjs.com/package/asciichart) for the inspiration.


CHANGELOG.md
# coinchart

## 1.1.3

### Patch Changes

- Fix unnecessary files uploaded to npm

## 1.1.2

### Patch Changes

- 444bc51: Fix incorrect bin dir, update readme

## 1.1.1

### Patch Changes

- 4393a6b: Fix missing token, update docs

## 1.1.0

### Minor Changes

- a4e001e: Moved project to typescript and refactored code


src/index.ts
#!/usr/bin/env node
import yargs from "yargs";
import { hideBin } from "yargs/helpers";
import axios from "axios";
import { CoinResponse, GraphResponse } from "./types";

const BASE_URL = "https://api.binance.com/api/v3";

const COLOURS = {
  green: "\u001b[32;1m",
  red: "\u001b[31;1m",
  yellow: "\u001b[33;1m",
  reset: "\u001b[0m",
};

const argv = yargs(hideBin(process.argv))
  .default({
    i: "1h",
    l: 50,
    p: "USDT",
    r: 25,
  })
  .help("h")
  .alias("h", "help")
  .alias("i", "interval")
  .describe("i", "Interval eg. 15m, 4h, 1d")
  .alias("p", "pair")
  .describe("p", "Coin pairing eg. BTC, ETH, BNB")
  .alias("l", "limit")
  .describe("l", "Number of candlesticks")
  .alias("r", "rows")
  .describe("r", "Graph height measured in rows")
  .parseSync();

const coin = (argv._.length ? argv._[0] : "BTC")?.toString().toUpperCase();
const { i: interval, l: limit, p: pairInput, r: graphHeight } = argv;
const pair = pairInput?.toUpperCase();

const coinDataUrl = `${BASE_URL}/ticker/24hr?symbol=${coin}${pair}`;
const graphDataUrl = `${BASE_URL}/klines?symbol=${coin}${pair}&interval=${interval}&limit=${limit}`;

// ==== Graph Response format ==== //
//
// [
//   [
//     1499040000000,      // Kline open time
//     "0.01634790",       // Open price
//     "0.80000000",       // High price
//     "0.01575800",       // Low price
//     "0.01577100",       // Close price
//     "148976.11427815",  // Volume
//     1499644799999,      // Kline Close time
//     "2434.19055334",    // Quote asset volume
//     308,                // Number of trades
//     "1756.87402397",    // Taker buy base asset volume
//     "28.46694368",      // Taker buy quote asset volume
//     "0"                 // Unused field, ignore.
//   ]
// ]
//
// Refer to https://github.com/binance/binance-spot-api-docs/blob/master/rest-api.md#klinecandlestick-data
//
// =============================== //

const fetcher = async <T>(url: string): Promise<T> =>
  axios.get(url, { timeout: 5000 }).then((res) => res.data);

(async () => {
  try {
    const [coinData, graphData] = await Promise.all([
      fetcher<CoinResponse>(coinDataUrl),
      fetcher<GraphResponse>(graphDataUrl),
    ]);

    if (!coinData || !graphData) {
      console.log("Error fetching data :( Please try again!");
      return;
    }

    const {
      openingPrices,
      closingPrices,
      highestPrices,
      lowestPrices,
      minPrice,
      maxPrice,
    } = graphData.reduce(
      (acc, d, i) => {
        const values = d.slice(1, 5).map((v) => parseFloat(v as string));

        acc.minPrice = Math.min(acc.minPrice, ...values);
        acc.maxPrice = Math.max(acc.maxPrice, ...values);

        const [open, high, low, close] = values;

        acc.openingPrices.push(open!);
        acc.highestPrices.push(high!);
        acc.lowestPrices.push(low!);
        acc.closingPrices.push(close!);

        return acc;
      },
      {
        openingPrices: [] as number[],
        closingPrices: [] as number[],
        highestPrices: [] as number[],
        lowestPrices: [] as number[],
        minPrice: parseFloat(graphData?.[0]?.[3] as string),
        maxPrice: parseFloat(graphData?.[0]?.[2] as string),
      }
    );

    const range = Math.abs(maxPrice - minPrice);

    const ratio = range !== 0 ? graphHeight / range : 1;

    // Recalculate min and max to fit graph height
    const adjustedMin = Math.round(minPrice * ratio);
    const adjustedMax = Math.round(maxPrice * ratio);

    const seriesLength = graphData.length;
    const labelOffset = 3;
    const height = Math.abs(adjustedMax - adjustedMin);
    const width = seriesLength + labelOffset;
    const padding = "         ";

    const graph = new Array(height + 1);
    for (let i = 0; i <= height; i++) {
      graph[i] = new Array(width);
      for (let j = 0; j < width; j++) {
        graph[i][j] = " ";
      }
    }

    // Draw y axis labels
    for (let val = adjustedMin; val <= adjustedMax; ++val) {
      const labelValue =
        height > 0 ? maxPrice - ((val - adjustedMin) * range) / height : val;

      const decimal = labelValue < 1 ? 4 : labelValue < 1000 ? 2 : 0;
      const label = (padding + labelValue.toFixed(decimal)).slice(
        -padding.length
      );
      graph[val - adjustedMin][Math.max(labelOffset - label.length, 0)] = label;
      graph[val - adjustedMin][labelOffset - 1] = "┤";
    }

    const getPriceYIndex = (price: number) =>
      Math.round(price * ratio) - adjustedMin;

    // Draw graph
    for (let x = 0; x < seriesLength; x++) {
      const openingPriceIndex = getPriceYIndex(openingPrices[x]!);
      const closingPriceIndex = getPriceYIndex(closingPrices[x]!);
      const highestPriceIndex = getPriceYIndex(highestPrices[x]!);
      const lowestPriceIndex = getPriceYIndex(lowestPrices[x]!);
      const currentColumn = x + labelOffset;
      const colour =
        openingPriceIndex <= closingPriceIndex ? COLOURS.green : COLOURS.red;

      for (
        let i = height - highestPriceIndex;
        i <= height - lowestPriceIndex;
        ++i
      ) {
        // Draw candle body if inbetween opening and closing price
        if (
          (i <= height - openingPriceIndex &&
            i >= height - closingPriceIndex) ||
          (i >= height - openingPriceIndex && i <= height - closingPriceIndex)
        ) {
          graph[i][currentColumn] = `${colour}█${COLOURS.reset}`;
        } else {
          graph[i][currentColumn] = `${colour}|${COLOURS.reset}`; // Draw candle wick if not inbetween opening and closing price
        }
      }
    }

    console.log(graph.map((x) => x.join(" ")).join("\n"));
    console.log(
      `\n ${COLOURS.yellow}${coin}/${pair}\u001b[0m  ${interval} chart`
    );

    const priceChangeColour =
      parseInt(coinData.priceChangePercent) >= 0 ? COLOURS.green : COLOURS.red;

    console.log(
      `\n The current price for ${coin} is ${priceChangeColour}${coinData.lastPrice} ${pair} (${coinData.priceChangePercent}%)`
    );
  } catch (error) {
    console.log("Oh no! An error occured. \n");

    if (axios.isAxiosError(error)) {
      console.error(`Error: ${error.message} \n`);
    }

    console.log("Please try again!");
  }
})();


index.js
#!/usr/bin/env node
const axios = require("axios");
const argv = require("yargs/yargs")(process.argv.slice(2))
	.default({
		i: "1h",
		l: 50,
		p: "USDT",
		r: 25,
	})
	.alias("v", "version")
	.help("h")
	.alias("h", "help")
	.alias("i", "interval")
	.describe("i", "Interval eg. 15m, 4h, 1d")
	.alias("p", "pair")
	.describe("p", "Coin pairing eg. BTC, BNB, ETH")
	.alias("l", "limit")
	.describe("l", "Number of candlesticks")
	.alias("r", "rows")
	.describe("r", "Graph height in rows").argv;

let [coin] = argv._.length ? argv._ : ["BTC"];
coin = coin.toUpperCase();
const interval = argv.i;
const limit = argv.l;
const pair = argv.p.toUpperCase();
const graphHeight = argv.r;
const green = "\u001b[32;1m";
const red = "\u001b[31;1m";
const stopColor = "\u001b[0m";

let coinData;
let color;

const getData = () => {
	const data = axios.get(`https://api.binance.com/api/v3/ticker/24hr?symbol=${coin}${pair}`);
	const graphData = axios.get(
		`https://api.binance.com/api/v3/klines?symbol=${coin}${pair}&interval=${interval}&limit=${limit}`
	);
	Promise.all([data, graphData])
		.then((res) => {
			coinData = res[0].data;
			color = res[0].data.priceChangePercent >= 0 ? green : red;
			plot(res[1].data);
		})
		.catch(() => console.log("Error fetching data :( Please try again!"));
};

const plot = (series) => {
	let seriesLength = series.length;

	let min = series[0][3];
	let max = series[0][2];

	const openArr = [];
	const closeArr = [];
	const highArr = [];
	const lowArr = [];

	series.forEach((d, i) => {
		let values = d.slice(1, 5);
		min = Math.min(min, ...values);
		max = Math.max(max, ...values);
		openArr.push(values[0]);
		highArr.push(values[1]);
		lowArr.push(values[2]);
		closeArr.push(values[3]);
	});

	let range = Math.abs(max - min);
	let height = graphHeight;
	let ratio = range !== 0 ? height / range : 1;

	let min2 = Math.round(min * ratio);
	let max2 = Math.round(max * ratio);

	let rows = Math.abs(max2 - min2);
	let width = seriesLength;
	let padding = "           ";
	let offset = 3;
	width = width + offset;

	let result = new Array(rows + 1);

	for (let i = 0; i <= rows; i++) {
		result[i] = new Array(width);
		for (let j = 0; j < width; j++) {
			result[i][j] = " ";
		}
	}

	const format = function (x) {
		const dec = x < 1 ? 4 : x < 1000 ? 2 : 0;
		return (padding + x.toFixed(dec)).slice(-padding.length);
	};

	for (let y = min2; y <= max2; ++y) {
		let label = format(rows > 0 ? max - ((y - min2) * range) / rows : y, y - min2);
		result[y - min2][Math.max(offset - label.length, 0)] = label;
		result[y - min2][offset - 1] = "┤";
	}

	for (let x = 0; x < seriesLength; x++) {
		let yOpen = Math.round(openArr[x] * ratio) - min2;
		let yClose = Math.round(closeArr[x] * ratio) - min2;
		let yHigh = Math.round(highArr[x] * ratio) - min2;
		let yLow = Math.round(lowArr[x] * ratio) - min2;

		const col = x + offset;

		if (yOpen <= yClose) {
			for (let i = rows - yHigh; i < rows - yClose; ++i) {
				result[i][col] = `${green}|${stopColor}`;
			}
			for (let i = rows - yOpen; i < rows - yLow; i++) {
				result[i][col] = `${green}|${stopColor}`;
			}
		} else {
			for (let i = rows - yOpen; i <= rows - yLow; ++i) {
				result[i][col] = `${red}|${stopColor}`;
			}
			for (let i = rows - yHigh; i <= rows - yClose; ++i) {
				result[i][col] = `${red}|${stopColor}`;
			}
		}

		//fill between open and close
		const low = Math.min(rows - yOpen, rows - yClose);
		const high = Math.max(rows - yOpen, rows - yClose);

		if (low !== high) {
			for (let i = low; i <= high; ++i) {
				result[i][col] = yOpen <= yClose ? `${green}█${stopColor}` : `${red}█${stopColor}`;
			}
		}
		result[rows - yOpen][col] =
			yOpen <= yClose ? `${green}█${stopColor}` : `${red}█${stopColor}`;
	}
	console.log(result.map((x) => x.join(" ")).join("\n"));
	console.log(`\n \u001b[1m\u001b[33;1m${coin}/${pair}\u001b[0m  ${interval} chart`);
	console.log(
		`\n The current price for ${coin} is \u001b[1m${color}${coinData.lastPrice} ${pair} (${coinData.priceChangePercent}%)`
	);
};
getData();
